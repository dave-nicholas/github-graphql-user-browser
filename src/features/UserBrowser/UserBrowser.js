import React from 'react';
import { debounce } from 'lodash';
import UserDetailsConnected from './UserDetails';

export default class UserBrowser extends React.Component {
  state = { username: 'gaearon', inputValue: 'gaearon' };

  onUsernameChange = inputValue => {
    this.setState({ inputValue });
    this.debouncedSetUsername(inputValue);
  };

  debouncedSetUsername = debounce(username => this.setState({ username }), 500);

  render() {
    return (
      <React.Fragment>
        <h1>Github User Browser</h1>
        <div className="input-field col s6">
          <input
            id="github_username"
            type="text"
            value={this.state.inputValue}
            onChange={e => this.onUsernameChange(e.target.value)}
          />
          <label htmlFor="github_username">Github Username</label>
        </div>
        <UserDetailsConnected username={this.state.username} />
      </React.Fragment>
    );
  }
}
