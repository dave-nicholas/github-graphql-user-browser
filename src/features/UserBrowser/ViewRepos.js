import React from 'react';
import PropTypes from 'prop-types';

const ViewRepos = ({ repositories: { edges } }) =>
  !!edges.length ? (
    <ul className="collection">
      {edges.map(({ node: { url, name, description } }, i) => (
        <li className="collection-item" key={i}>
          <span className="title">
            <a href={url}>{name}</a>
          </span>
          <p>{description}</p>
        </li>
      ))}
    </ul>
  ) : (
    'No repos...'
  );

ViewRepos.propTypes = {
  repositories: PropTypes.shape({
    edges: PropTypes.arrayOf(
      PropTypes.shape({
        url: PropTypes.string,
        name: PropTypes.string,
        description: PropTypes.string
      })
    )
  }).isRequired
};

export default ViewRepos;
