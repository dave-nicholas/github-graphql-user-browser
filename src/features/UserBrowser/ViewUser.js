import React from 'react';
import PropTypes from 'prop-types';
import moment from 'moment';

const ViewUser = ({ user, avatarUrl, createdAt }) => {
  return (
    <div className="card">
      <div className="card-image">
        <img src={avatarUrl} alt={user} />
      </div>
      <div className="card-stacked">
        <div className="card-content">
          User since: {moment(createdAt).format('MMMM Do YYYY')}
        </div>
      </div>
    </div>
  );
};

ViewUser.propTypes = {
  avatarUrl: PropTypes.string.isRequired,
  createdAt: PropTypes.string.isRequired,
  user: PropTypes.string.isRequired
};

export default ViewUser;
