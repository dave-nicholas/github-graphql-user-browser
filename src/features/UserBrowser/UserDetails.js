import React from 'react';
import PropTypes from 'prop-types';
import { graphql } from 'react-apollo';
import gql from 'graphql-tag';
import Loader from '../../components/Loader';
import ViewRepos from './ViewRepos';
import ViewUser from './ViewUser';

export const UserDetails = ({ data: { user, loading }, username }) => {
  if (loading) return <Loader />;
  if (!user) return 'User not found.';
  const { avatarUrl, createdAt, repositories } = user;
  return (
    <div className="row">
      <div className="col s12 m6">
        <ViewUser avatarUrl={avatarUrl} user={username} createdAt={createdAt} />
      </div>
      <div className="col s12 m6">
        <ViewRepos repositories={repositories} />
      </div>
    </div>
  );
};

export const UserDetailQuery = gql`
  query($username: String!) {
    user(login: $username) {
      avatarUrl
      createdAt
      repositories(last: 5) {
        edges {
          node {
            url
            name
            description
          }
        }
      }
    }
  }
`;

export const UserDetailsWrapper = graphql(UserDetailQuery, {
  options: ({ username }) => ({ variables: { username }, errorPolicy: 'all' })
});

const UserDetailsConnected = UserDetailsWrapper(UserDetails);

UserDetailsConnected.propTypes = {
  username: PropTypes.string.isRequired
};

export default UserDetailsConnected;
