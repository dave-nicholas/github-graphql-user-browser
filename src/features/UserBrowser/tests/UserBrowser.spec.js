import React from 'react';
import { shallow } from 'enzyme';
import UserBrowser from '../UserBrowser';
import toJson from 'enzyme-to-json';

describe('UserBrowser', () => {
  it('should call debouncedSetUsername', () => {
    const wrapper = shallow(<UserBrowser />);

    wrapper.instance().debouncedSetUsername = jest.fn();

    wrapper.instance().onUsernameChange('test');
    expect(wrapper.instance().debouncedSetUsername).toHaveBeenCalledWith(
      'test'
    );
  });

  it('should match snapshot', () => {
    const wrapper = shallow(<UserBrowser />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
