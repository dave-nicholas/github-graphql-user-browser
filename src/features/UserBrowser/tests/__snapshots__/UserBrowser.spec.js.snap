// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserBrowser should match snapshot 1`] = `
<React.Fragment>
  <h1>
    Github User Browser
  </h1>
  <div
    className="input-field col s6"
  >
    <input
      id="github_username"
      onChange={[Function]}
      type="text"
      value="gaearon"
    />
    <label
      htmlFor="github_username"
    >
      Github Username
    </label>
  </div>
  <Apollo(Component)
    username="gaearon"
  />
</React.Fragment>
`;
